onnxruntime/core/optimizer/double_qdq_pairs_remover.cc

```diff
@@ -0,0 +1,182 @@
+  const ONNX_NAMESPACE::TensorProto* node2_scale_tensor_proto =
+      graph_utils::GetConstantInitializer(graph, node2.InputDefs()[InputIndex::SCALE_ID]->Name());
+  const ONNX_NAMESPACE::TensorProto* node1_zp_tensor_proto =
+      graph_utils::GetConstantInitializer(graph, node1.InputDefs()[InputIndex::ZERO_POINT_ID]->Name());
+  const ONNX_NAMESPACE::TensorProto* node2_zp_tensor_proto =
+      graph_utils::GetConstantInitializer(graph, node2.InputDefs()[InputIndex::ZERO_POINT_ID]->Name());
+  Initializer zero_point_init_1{*node1_zp_tensor_proto, graph.ModelPath()};
+  Initializer zero_point_init_2{*node2_zp_tensor_proto, graph.ModelPath()};
+  if (zero_point_init_1.data_type() != zero_point_init_2.data_type()) {
+    return false;
+  }
+  Initializer scale_init_1{*node1_scale_tensor_proto, graph.ModelPath()};
+  Initializer scale_init_2{*node2_scale_tensor_proto, graph.ModelPath()};
+  if (scale_init_1.data_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT ||
+      scale_init_2.data_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+    return false;
+  }
+  const float scale_1 = scale_init_1.data<float>()[0];
+  const float scale_2 = scale_init_2.data<float>()[0];
+
+  float real_max1 = 0.0;
+  float real_min1 = 0.0;
+  float real_max2 = 0.0;
+  float real_min2 = 0.0;
+  int zero_point_1 = 0;
+  int zero_point_2 = 0;
+  if (zero_point_init_1.data_type() == ONNX_NAMESPACE::TensorProto_DataType_UINT8) {
+    zero_point_1 = zero_point_init_1.data<uint8_t>()[0];
+  } else if (zero_point_init_1.data_type() == ONNX_NAMESPACE::TensorProto_DataType_INT8) {
+    zero_point_1 = zero_point_init_1.data<int8_t>()[0];
+  } else {
+    return false;
+  }
+  if (zero_point_init_2.data_type() == ONNX_NAMESPACE::TensorProto_DataType_UINT8) {
+    zero_point_2 = zero_point_init_2.data<uint8_t>()[0];
+  } else if (zero_point_init_2.data_type() == ONNX_NAMESPACE::TensorProto_DataType_INT8) {
+    zero_point_2 = zero_point_init_2.data<int8_t>()[0];
+  } else {
+    return false;
+  }
+  const int q_min_1 = (zero_point_init_1.data_type() == ONNX_NAMESPACE::TensorProto_DataType_UINT8) ? std::numeric_limits<uint8_t>::min() : std::numeric_limits<int8_t>::min();
+  const int q_max_1 = (zero_point_init_1.data_type() == ONNX_NAMESPACE::TensorProto_DataType_UINT8) ? std::numeric_limits<uint8_t>::max() : std::numeric_limits<int8_t>::max();
+  const int q_min_2 = (zero_point_init_2.data_type() == ONNX_NAMESPACE::TensorProto_DataType_UINT8) ? std::numeric_limits<uint8_t>::min() : std::numeric_limits<int8_t>::min();
+  const int q_max_2 = (zero_point_init_2.data_type() == ONNX_NAMESPACE::TensorProto_DataType_UINT8) ? std::numeric_limits<uint8_t>::max() : std::numeric_limits<int8_t>::max();
+
+  real_min1 = q_min_1 - zero_point_1 * scale_1;
+  real_min2 = q_min_2 - zero_point_2 * scale_2;
+  real_max1 = (q_max_1 - q_min_1 + zero_point_1) * scale_1 - q_min_1;
+  real_max2 = (q_max_2 - q_min_2 + zero_point_2) * scale_2 - q_min_1;
```

Comment:
This computation is not right. The formula is real_value_fp32 = scale * ( Integer_value - zero_point), so
real_min1 = scale_1 * ( q_min_1 - zero_point_1)
real_max1 = scale_1 * ( q_max_1 - zero_point_1). 
Similar to the 2nd node

Reply:
Resolved